_posts/2021-04-09-terraform-aws.md

```diff
@@ -0,0 +1,494 @@
+  validate      Check whether the configuration is valid
+  plan          Show changes required by the current configuration
+  apply         Create or update infrastructure
+  destroy       Destroy previously-created infrastructure
+
+All other commands:
+  console       Try Terraform expressions at an interactive command prompt
+  fmt        
+...
+...
+```
+We will use the main commands `init`, `plan`, and `apply` throughout this post.
+
+Since we will be creating resources in AWS, we will also set up the AWS CLI by running the below command:
+```shell
+aws configure
+```
+
+When prompted, we will provide AWS access key id and secret access key and choose a default region and output format:
+
+```shell
+AWS Access Key ID [****************2345]: ....
+AWS Secret Access Key [****************2345]: ...
+Default region name [us-east-1]: 
+Default output format [json]: 
+```
+We are using us-east-1 as the region and JSON as the output format.
+
+## Terraform Concepts with Simple Workflow
+For a basic workflow in Terraform, we first design the infrastructure resources in a configuration file, called the desired state. We then use this configuration to create the actual infrastructure.
+
+The configuration is defined in [Terraform language](https://www.terraform.io/docs/language/index.html) using a JSON-like syntax called Hashicorp Configuration Language(HCL) that tells Terraform how to manage a collection of infrastructure. A configuration can consist of one or more files and directories.
+
+### Init, Plan, and Apply Cycle
+We start with our desired state that is the collection of infrastructure resources we wish to create. When we run the `plan`, Terraform pulls the actual resource information from the provider and compares it with the desired infrastructure. It then outputs a report containing the changes which will happen when the configuration is applied(during the `Apply` stage).
+
+The main steps for any basic task with Terraform are:
+1. Create a workspace folder
+2. Create a configuration file named `main.tf`. This file will have the configuration of the infrastructure to be created or updated.
+3. Initialize the workspace using the command `terraform init`.
+4. Create the plan
+5. Apply the plan
+
+### Defining the Configuration
+Let us define our Terraform configuration in Terraform language in a file `main.tf`:
+
+```
+terraform {
+  required_providers {
+    aws = {
+      source  = "hashicorp/aws"
+      version = "~> 3.27"
+    }
+  }
+}
+
+provider "aws" {
+  profile = "default"
+  region  = "us-west-2"
+}
+
+resource "aws_instance" "vm-web" {
+  ami           = "ami-830c94e3"
+  instance_type = "t2.micro"
+
+  tags = {
+    Name = "server for web"
+    Env = "dev"
+  }
+}
+```
+Here we are creating an aws EC2 instance named "vm-web" of type `t2.micro` using an ami `ami-830c94e3`. We also define two tags with names - `Name` and `Env` for associating with ec2 instance.
+
+We can also see the three main parts of configuration :
```

Comment:
Can you maybe add a link to each point in the list below so the readers can see where they can find documentation about resources, providers, and terraform settings?